### PR TITLE
Mappers

### PR DESCRIPTION
**Pull request by cryogenian**
[ch5612]

+ Abstracted failing cases into `ApplicativePlus`
+ Abstracted unique key (wrapping uuid) into state
+ Added mapper to state
+ Modified pipes to be mapped 

`Projection` is mappable thing, it is changed when we focus or unfocus.
I switched to `O.string("$" concat whatever)` for temporary projections, other option is to create a case class that wraps such one level non-mappable projection, but I don't see any pros in it. 

Oh! I also switched to scalaz almost everywhere mostly because of matryoshka depending on it and I wanted compiler working well with interpreter.

### CODE REVIEW

**Comment by rintcius** [comment]
@wemrysi and @jsantos17 - would also be good if you have a look at this PR

**Comment by rintcius** [comment]
I did a quick scan on `$replaceRoot` but didn't see this. Wasn't this a purpose of this `Mapper`?

**Comment by cryogenian** [comment]
Nah, `$replaceRoot` isn't used here. It's tempting to add it somewhere, but there is still no case for it. 

Initially mongo ds produced all records wrapped with uuid, in some situation this isn't necessary (e.g. when `Wrap` or `Cartesian` is the last stage), and so we don't need to wrap --> no need in `BsonDocument` map. Basically this pr is about what @wemrysi said, we don't need to construct wrapper when we don't need to 🙂 

I also considered using map from one field to another to avoid situation when we can't interpret `Cartesian` due to have dotty aliases, but it seems that making thoth returning aliases w/o dots _much_ simpler.

**Comment by cryogenian** [comment]
[ch5870] [ch5871]

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala [hunk 1]
```diff
@@ -82,7 +99,8 @@ class MongoScalarStagesInterpreterSpec
         (CPathField("a1"), (CPathField("a0"), Nil)),
         (CPathField("b1"), (CPathField("b0"), Nil)))
 
-      input must cartesianInto(targets)(expected)
+      val actual = interpret(ScalarStages(IdStatus.ExcludeId, List(ScalarStage.Cartesian(targets))), input, (x => x))
+      actual must bestSemanticEqual(expected)
     }
 
     "nested pivoting doesn't produce unnecessary empty fields" in {
```
Comment (alissapajer) [inline datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: This test is `cart-7` in `ScalarStageSpec`. Can we delete it here?

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala [hunk 2]
```diff
@@ -66,12 +66,29 @@ class MongoScalarStagesInterpreterSpec
     }
   }
 
+  "Pivot special" >> {
+    "no unnecessary undefined pairs" in {
```
Comment (alissapajer) [inline datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: I'm in the process of adding more `IncludeId` and `IdOnly` pivot tests in `ScalarStagesSpec`. I'll add this one there, and then we can delete it here. In general, we should put all the tests we can in `ScalarStagesSpec` and only put tests specific to individual implementations here.

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala [hunk 3]
```diff
@@ -66,12 +66,29 @@ class MongoScalarStagesInterpreterSpec
...
+        ["a", 1]
+        ["b", 2]
+        """)
+      val targets = Pivot(IdStatus.IncludeId, ColumnType.Object)
+      evalPivot(targets, input) must bestSemanticEqual(expected)
+    }
+  }
+
   "Cartesian special" >> {
     "cross fields when some are undefined" in {
```
Comment (alissapajer) [inline datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: I'll move this one over to `ScalarStagesSpec` as well. We don't need to hold up merge of this PR though.

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala [hunk 4]
```diff
@@ -27,16 +27,16 @@ import quasar.{IdStatus, ScalarStageSpec => Spec, ScalarStage, ScalarStages}
...
     with Spec.CartesianSpec
     with StageInterpreterSpec {
-
-  val idsPendingExamples: Set[Int] = Set()
   val wrapPendingExamples: Set[Int] = Set()
   val projectPendingExamples: Set[Int] = Set()
   val maskPendingExamples: Set[Int] = Set()
-  val pivotPendingExamples: Set[Int] = Set(9, 10, 11)
+  val pivotPendingExamples: Set[Int] = Set()
+  val focusedPendingExamples: Set[Int] = Set()
```
Comment (alissapajer) [inline datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: :tada:

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala (hunk 1 above)
Comment (cryogenian) [reply datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: Absolutely!

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala (hunk 3 above)
Comment (cryogenian) [reply datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: This is very specific test actually, during development I got `["whatever_pivot_undefined", "whatever_pivot_undefined"]`, I don't think that this is common case for all interpreters

**Comment by djspiewak** [comment]
> Oh! I also switched to scalaz almost everywhere mostly because of matryoshka depending on it and I wanted compiler working well with interpreter.

I don't have any objections to this at all, you should always use what makes the most sense, but it is worth noting that `import shims._` should make the differences between the two broadly meaningless. Matryoshka should work just as well with cats (when shims is imported) as it works with scalaz. If it doesn't, I'd like to know about it so I can fix shims. :-)

**Comment by cryogenian** [comment]
@djspiewak Shims work fine, but I had drastically longer compile times before switched to scalaz (like 1s -> 10s for one module recompilation)

**Comment by djspiewak** [comment]
@cryogenian Yeah that makes sense. There are a lot of places where, if the type is unfixed, Scala can figure it out through implicit solution. Shims complicates this considerably because it vastly increases the space of possible implicits.

In truth, we shouldn't be relying on implicit solution in most of those cases, since it's *always* going to ultimately lead to enormously long compile times whenever the file is even a little bit larger, or when the type becomes more complex, but shims undoubtedly exacerbates the issues.

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala (hunk 3 above)
Comment (alissapajer) [reply datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: If there is a mongo-specific reason to test this here, then :+1: Maybe add a comment here about those undefined values so we know why this test is important.

File: datasource/src/main/scala/quasar/physical/mongo/expression/Pipeline.scala [hunk 5]
```diff
@@ -29,28 +29,45 @@ trait MongoPipeline[+A] extends Pipeline[A]
...
+          NonTerminal(List("MaskFilter"), None, List(fld.render))
+      }
   }
 
-  final case class NotNull(field: String) extends CustomPipeline
+  final case class MaskFilter(field: String) extends CustomPipeline
+  final case class PivotFilter(field: String) extends CustomPipeline
 
-  def pipeMinVersion[A](pipe: MongoPipeline[A]): Version = pipe match {
+  def pipeMinVersion(pipe: MongoPipeline[_]): Version = pipe match {
```
[the remark this answers is not in the document]
Comment (cryogenian) [reply datasource/src/main/scala/quasar/physical/mongo/expression/Pipeline.scala]: It communicates more clear that this is existential not universal. I.e. no matter what's inside, version can be grabbed

**Comment by jsantos17** [comment]
Released as `quasar-datasource-mongo_2.12-4.2.4-c26fdc7`

**Comment by cryogenian** [comment]
@rintcius 

We definitely haven't reached stability yet. I _really_ dislike it but any change in interpreters need to change their tests (mostly because they are not tests, but examples showing how pipeline is compiled). 

E.g. to fix foc-11 I needed to change `Mask` interpreter, this means that some of examples now broken, so, I need to change examples too. And this is absolutely synthetic and mechanic change because `Mask` interpreter output is so complicated, i.e. I can predict changes in example and check that it's compiled down to correct `CoreOp`, but writing it by hand is tedious and error-prone 🙁

**Comment by cryogenian** [comment]
@rintcius 
I'll close this pr tomorrow and split it into two prs: 
+ One is what it used to be before test fixes
+ The second one based on the first but with `_id` filters moved from initial pipes to interperters

**Comment by alissapajer** [comment]
@cryogenian @jsantos17 I minimized mongo's bug that was exposed in `multilevelFlatten.test`. The problem exists with arrays but not with objects, as exhibited by the first two tests. The third test is failing due to the same reason, most likely (it is a partial minimization of the original issue). 

I'm going to add the first two tests, and maybe the third, to `ScalarStageSpec` since they aren't special to mongo. I put there here so you can work off them in the interim.
```
[info] tests
[error]   x disjunctively retain Mask values in an array
[error]    List(BsonArray{values=[BsonString{value='c'}, { }]}, BsonArray{values=[BsonString{value='e'}, { "z" : { "$numberLong" : "14" } }]})
[error]      is missing: BsonArray{values=[BsonString{value='a'}]}, BsonArray{values=[BsonString{value='b'}]}, BsonArray{values=[BsonString{value='d'}]} (MongoScalarStagesInterpreterSpec.scala:124)
[info]   + disjunctively retain Mask values in an object
[error]   x disjunctively retain Mask values in an array after pivoting
[error]    List(BsonArray{values=[BsonString{value='midObj'}, { "bot" : false }]})
[error]      is missing: BsonArray{values=[BsonString{value='midArr'}]} (MongoScalarStagesInterpreterSpec.scala:163)
```

```scala
  "tests" >> {
    "disjunctively retain Mask values in an array" in {
      val input = ldjson("""
        ["a", 13]
        ["b", []]
        ["c", {}]
        ["d", [12]]
        ["e", { "z": 14}]
        """)
      val expected = ldjson("""
        ["a"]
        ["b"]
        ["c", {}]
        ["d"]
        ["e", { "z": 14}]
        """)

      input must maskInto("[0]" -> ColumnType.Top, "[1]" -> Set(ColumnType.Object))(expected)
    }   

    "disjunctively retain Mask values in an object" in {
      val input = ldjson("""
        { "v": "a", "w": 13 }
        { "v": "b", "w": [] }
        { "v": "c", "w": {} }
        { "v": "d", "w": [12] }
        { "v": "e", "w": { "z": 14} }
        """)
      val expected = ldjson("""
        { "v": "a" }
        { "v": "b" }
        { "v": "c", "w": {} }
        { "v": "d" }
        { "v": "e", "w": { "z": 14} }
        """)

      input must maskInto(".v" -> ColumnType.Top, ".w" -> Set(ColumnType.Object))(expected)
    }   

    "disjunctively retain Mask values in an array after pivoting" in {
      val input = ldjson("""
        { "topObj": { "midArr":[12], "midObj": { "bot":false } } }
        """)
      val expected = ldjson("""
        ["midArr"]
        ["midObj", { "bot": false}]
        """)

      val stages = List(
        Project(CPath.parse("topObj")),
        Mask(Map(CPath.Identity -> Set(ColumnType.Object))),
        Pivot(IdStatus.IncludeId, ColumnType.Object),
        Mask(Map(
          CPath.parse("[0]") -> ColumnType.Top,
          CPath.parse("[1]") -> Set(ColumnType.Object))))

      input must interpretInto(stages)(expected)
    }   
  }
```

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala (hunk 3 above)
Comment (alissapajer) [reply datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: I was thinking more actually, and I'd like to have this test in `ScalarStageSpec`. Even though mongo hit this case for reasons specific to the mongo implementation, it's still a valid test for all interpreters. And the more tests that we can have over in `ScalarStageSpec` the better, because you never know what corner cases different impls will have.

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala (hunk 2 above)
Comment (alissapajer) [reply datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: Added here https://github.com/slamdata/quasar/pull/4101/files#diff-018b1055badf3c3df25b2b218bb1abcbR1161

File: datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala (hunk 3 above)
Comment (alissapajer) [reply datasource/src/test/scala/quasar/physical/mongo/MongoScalarStagesInterpreterSpec.scala]: Added here https://github.com/slamdata/quasar/pull/4101/files#diff-018b1055badf3c3df25b2b218bb1abcbR2162

**Comment by alissapajer** [comment]
Added the `Mask` tests above starting here https://github.com/slamdata/quasar/pull/4101/files#diff-018b1055badf3c3df25b2b218bb1abcbR819 and a `Pivot . Mask` test here https://github.com/slamdata/quasar/pull/4101/files#diff-018b1055badf3c3df25b2b218bb1abcbR1760

**Comment by alissapajer** [comment]
Ok, against quasar-version `20.0.0-a59904a` (slamdata/quasar#4101) the following tests are failing:

`mask-23`, `mask-24`, `foc-23` (the tests from above)

`pivot-16`, `pivot-28`, `pivot-31` (looks like a bug with `IdOnly` and array pivoting)

File: datasource/src/main/scala/quasar/physical/mongo/expression/Pipeline.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (rintcius) [reply datasource/src/main/scala/quasar/physical/mongo/expression/Pipeline.scala]: Oh, for me it communicates that something didn't work out with type inference ;)

[the remark this answers is not in the document]
Comment (cryogenian) [reply datasource/src/main/scala/quasar/physical/mongo/expression/Pipeline.scala]: That's true too 😄

**Comment by cryogenian** [comment]
Closing in favour of #114
